File: tests/algorithms/transpiler/FS/sorts/counting_sort.fs
// Generated 2025-08-11 16:20 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let _arrset (arr:'a array) (i:int) (v:'a) : 'a array =
    let mutable a = arr
    if i >= a.Length then
        let na = Array.zeroCreate<'a> (i + 1)
        Array.blit a 0 na 0 a.Length
        a <- na
    a.[i] <- v
    a
let rec _str v =
    let s = sprintf "%A" v
    let s = if s.EndsWith(".0") then s.Substring(0, s.Length - 2) else s
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec max_val (arr: int array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable arr = arr
    try
        let mutable m: int = _idx arr (int 0)
        let mutable i: int = 1
        while i < (Seq.length (arr)) do
            if (_idx arr (int i)) > m then
                m <- _idx arr (int i)
            i <- i + 1
        __ret <- m
        raise Return
        __ret
    with
        | Return -> __ret
let rec min_val (arr: int array) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable arr = arr
    try
        let mutable m: int = _idx arr (int 0)
        let mutable i: int = 1
        while i < (Seq.length (arr)) do
            if (_idx arr (int i)) < m then
                m <- _idx arr (int i)
            i <- i + 1
        __ret <- m
        raise Return
        __ret
    with
        | Return -> __ret
let rec counting_sort (collection: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable collection = collection
    try
        if (Seq.length (collection)) = 0 then
            __ret <- Array.empty<int>
            raise Return
        let coll_len: int = Seq.length (collection)
        let coll_max: int = max_val (collection)
        let coll_min: int = min_val (collection)
        let counting_arr_length: int = (coll_max + 1) - coll_min
        let mutable counting_arr: int array = Array.empty<int>
        let mutable i: int = 0
        while i < counting_arr_length do
            counting_arr <- Array.append counting_arr [|0|]
            i <- i + 1
        i <- 0
        while i < coll_len do
            let number: int = _idx collection (int i)
            counting_arr.[int (number - coll_min)] <- (_idx counting_arr (int (number - coll_min))) + 1
            i <- i + 1
        i <- 1
        while i < counting_arr_length do
            counting_arr.[int i] <- (_idx counting_arr (int i)) + (_idx counting_arr (int (i - 1)))
            i <- i + 1
        let mutable ordered: int array = Array.empty<int>
        i <- 0
        while i < coll_len do
            ordered <- Array.append ordered [|0|]
            i <- i + 1
        let mutable idx: int = coll_len - 1
        while idx >= 0 do
            let number: int = _idx collection (int idx)
            let pos: int = (_idx counting_arr (int (number - coll_min))) - 1
            ordered.[int pos] <- number
            counting_arr.[int (number - coll_min)] <- (_idx counting_arr (int (number - coll_min))) - 1
            idx <- idx - 1
        __ret <- ordered
        raise Return
        __ret
    with
        | Return -> __ret
let ascii_chars: string = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~"
let rec chr (code: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable code = code
    try
        if code = 10 then
            __ret <- "\n"
            raise Return
        if code = 13 then
            __ret <- "\r"
            raise Return
        if code = 9 then
            __ret <- "\t"
            raise Return
        if (code >= 32) && (code < 127) then
            __ret <- _substring ascii_chars (code - 32) (code - 31)
            raise Return
        __ret <- ""
        raise Return
        __ret
    with
        | Return -> __ret
let rec ord (ch: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable ch = ch
    try
        if ch = "\n" then
            __ret <- 10
            raise Return
        if ch = "\r" then
            __ret <- 13
            raise Return
        if ch = "\t" then
            __ret <- 9
            raise Return
        let mutable i: int = 0
        while i < (String.length (ascii_chars)) do
            if (_substring ascii_chars (i) (i + 1)) = ch then
                __ret <- 32 + i
                raise Return
            i <- i + 1
        __ret <- 0
        raise Return
        __ret
    with
        | Return -> __ret
let rec counting_sort_string (s: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable s = s
    try
        let mutable codes: int array = Array.empty<int>
        let mutable i: int = 0
        while i < (String.length (s)) do
            codes <- Array.append codes [|(ord (_substring s (i) (i + 1)))|]
            i <- i + 1
        let sorted_codes: int array = counting_sort (codes)
        let mutable res: string = ""
        i <- 0
        while i < (Seq.length (sorted_codes)) do
            res <- res + (chr (_idx sorted_codes (int i)))
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let example1: int array = counting_sort (unbox<int array> [|0; 5; 3; 2; 2|])
printfn "%s" (_str (example1))
let example2: int array = counting_sort (Array.empty<int>)
printfn "%s" (_str (example2))
let example3: int array = counting_sort (unbox<int array> [|-2; -5; -45|])
printfn "%s" (_str (example3))
printfn "%s" (counting_sort_string ("thisisthestring"))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/PHP/sorts/counting_sort.php
<?php
ini_set('memory_limit', '-1');
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function _iadd($a, $b) {
    if (function_exists('bcadd')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcadd($sa, $sb, 0);
    }
    return $a + $b;
}
function _isub($a, $b) {
    if (function_exists('bcsub')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcsub($sa, $sb, 0);
    }
    return $a - $b;
}
function _imul($a, $b) {
    if (function_exists('bcmul')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return bcmul($sa, $sb, 0);
    }
    return $a * $b;
}
function _idiv($a, $b) {
    return _intdiv($a, $b);
}
function _imod($a, $b) {
    if (function_exists('bcmod')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcmod($sa, $sb));
    }
    return $a % $b;
}
function max_val($arr) {
  global $ascii_chars, $example1, $example2, $example3;
  $m = $arr[0];
  $i = 1;
  while ($i < count($arr)) {
  if ($arr[$i] > $m) {
  $m = $arr[$i];
}
  $i = _iadd($i, 1);
};
  return $m;
}
function min_val($arr) {
  global $ascii_chars, $example1, $example2, $example3;
  $m = $arr[0];
  $i = 1;
  while ($i < count($arr)) {
  if ($arr[$i] < $m) {
  $m = $arr[$i];
}
  $i = _iadd($i, 1);
};
  return $m;
}
function counting_sort($collection) {
  global $ascii_chars, $example1, $example2, $example3;
  if (count($collection) == 0) {
  return [];
}
  $coll_len = count($collection);
  $coll_max = max_val($collection);
  $coll_min = min_val($collection);
  $counting_arr_length = _isub(_iadd($coll_max, 1), $coll_min);
  $counting_arr = [];
  $i = 0;
  while ($i < $counting_arr_length) {
  $counting_arr = _append($counting_arr, 0);
  $i = _iadd($i, 1);
};
  $i = 0;
  while ($i < $coll_len) {
  $number = $collection[$i];
  $counting_arr[_isub($number, $coll_min)] = _iadd($counting_arr[_isub($number, $coll_min)], 1);
  $i = _iadd($i, 1);
};
  $i = 1;
  while ($i < $counting_arr_length) {
  $counting_arr[$i] = _iadd($counting_arr[$i], $counting_arr[_isub($i, 1)]);
  $i = _iadd($i, 1);
};
  $ordered = [];
  $i = 0;
  while ($i < $coll_len) {
  $ordered = _append($ordered, 0);
  $i = _iadd($i, 1);
};
  $idx = _isub($coll_len, 1);
  while ($idx >= 0) {
  $number = $collection[$idx];
  $pos = _isub($counting_arr[_isub($number, $coll_min)], 1);
  $ordered[$pos] = $number;
  $counting_arr[_isub($number, $coll_min)] = _isub($counting_arr[_isub($number, $coll_min)], 1);
  $idx = _isub($idx, 1);
};
  return $ordered;
}
$ascii_chars = ' !"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~';
function mochi_chr($code) {
  global $ascii_chars, $example1, $example2, $example3;
  if ($code == 10) {
  return '
';
}
  if ($code == 13) {
}
function mochi_ord($ch) {
}
}
function counting_sort_string($s) {
}
$example1 = counting_sort([0, 5, 3, 2, 2]);
echo rtrim(_str($example1)), PHP_EOL;
$example2 = counting_sort([]);
echo rtrim(_str($example2)), PHP_EOL;
$example3 = counting_sort([-2, -5, -45]);
echo rtrim(_str($example3)), PHP_EOL;
echo rtrim(counting_sort_string('thisisthestring')), PHP_EOL;
  if ($ch == '	') {
  return 9;
}
  $i = 0;
  while ($i < strlen($ascii_chars)) {
  if (substr($ascii_chars, $i, _iadd($i, 1) - $i) == $ch) {
  return _iadd(32, $i);
}
  $i = _iadd($i, 1);
};
  return 0;
};
  function counting_sort_string($s) {
  global $ascii_chars, $example1, $example2, $example3;
  $codes = [];
  $i = 0;
  while ($i < strlen($s)) {
  $codes = _append($codes, mochi_ord($s[$i]));
  $i = _iadd($i, 1);
};
  $sorted_codes = counting_sort($codes);
  $res = '';
  $i = 0;
  while ($i < count($sorted_codes)) {
  $res = $res . mochi_chr($sorted_codes[$i]);
  $i = _iadd($i, 1);
};
  return $res;
};
  $example1 = counting_sort([0, 5, 3, 2, 2]);
  echo rtrim(_str($example1)), PHP_EOL;
  $example2 = counting_sort([]);
  echo rtrim(_str($example2)), PHP_EOL;
  $example3 = counting_sort([-2, -5, -45]);
  echo rtrim(_str($example3)), PHP_EOL;
  echo rtrim(counting_sort_string('thisisthestring')), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
